File: C++/src/cdVector.cpp
#include "cdVector.h"

MTypeId cdVector::id(0x123e1);

MObject cdVector::startMat;   //World Matrix of our starting obj.
MObject cdVector::endMat;     //World Matrix of our ending obj.
MObject cdVector::mag;        //Magnitude of vector.
MObject cdVector::normAttr;   //Normalize boolean.
MObject cdVector::outVec;     //Output Vector.
MObject cdVector::ox;         //X value of vector.
MObject cdVector::oy;         //Y value of vector.
MObject cdVector::oz;         //Z value of vector.

cdVector::cdVector() 
{
}

cdVector::~cdVector() 
{
}

void* cdVector::creator()
{ 
    return new cdVector(); 
}


MStatus cdVector::compute(const MPlug& plug, MDataBlock& data)
{

    if((plug == mag)||(plug == outVec)||(plug == ox)||(plug == oy)||(plug == oz))
    {

        MStatus status;

        MMatrix startMatrix = data.inputValue( startMat, &status ).asMatrix();
        CHECK_MSTATUS_AND_RETURN_IT( status );

        MMatrix endMatrix = data.inputValue( endMat, &status ).asMatrix();
        CHECK_MSTATUS_AND_RETURN_IT( status );

        bool norm = data.inputValue( normAttr, &status ).asBool();
        CHECK_MSTATUS_AND_RETURN_IT( status );

        MVector startVec(startMatrix[3][0], startMatrix[3][1], startMatrix[3][2]);
        MVector endVec(endMatrix[3][0], endMatrix[3][1], endMatrix[3][2]);

        MVector upVec = endVec - startVec;
        float dist = upVec.length();

        //Set the outgoing plugs.
        MDataHandle outVecX = data.outputValue(ox);
        MDataHandle outVecY = data.outputValue(oy);
        MDataHandle outVecZ = data.outputValue(oz);
        MDataHandle magVal = data.outputValue(mag);

        if (dist > 0.001f )
        {

            if (norm == true)
            {
                upVec.normalize();
                outVecX.setFloat(upVec.x);
                outVecY.setFloat(upVec.y);
                outVecZ.setFloat(upVec.z);
                magVal.setFloat(1.0f);
            }

            else
            {
                outVecX.setFloat(upVec.x);
                outVecY.setFloat(upVec.y);
                outVecZ.setFloat(upVec.z);
                magVal.setFloat(dist);
            }

        }

        else
        {
            outVecX.setFloat(0.0f);
            outVecY.setFloat(0.0f);
            outVecZ.setFloat(0.0f);
        }

        data.setClean(plug);

    }

    else
    {
        return MS::kUnknownParameter;
    }

    return MS::kSuccess;

}

MStatus cdVector::initialize()
{
    MStatus status;

    MFnNumericAttribute nAttr;
    MFnMatrixAttribute mAttr;

    startMat = mAttr.create("startMatrix", "startMat", MFnMatrixAttribute::kDouble);
    addAttribute(startMat);

    endMat = mAttr.create("endMatrix", "endMat", MFnMatrixAttribute::kDouble);
    addAttribute(endMat);

    mag = nAttr.create("endPoint", "ePoint", MFnNumericData::kFloat);
    addAttribute(mag);

    normAttr = nAttr.create("normalize", "norm", MFnNumericData::kBoolean);
    addAttribute(normAttr);

    ox = nAttr.create("outVectorX", "ox", MFnNumericData::kFloat, 0.0);
    nAttr.setWritable(false);
    nAttr.setStorable(true);
    nAttr.setReadable(true);

    oy = nAttr.create("outVectorY", "oy", MFnNumericData::kFloat, 0.0);
    nAttr.setWritable(false);
    nAttr.setStorable(true);
    nAttr.setReadable(true);

    oz = nAttr.create("outVectorZ", "oz", MFnNumericData::kFloat, 0.0);
    nAttr.setWritable(false);
    nAttr.setStorable(true);
    nAttr.setReadable(true);

    outVec = nAttr.create("outputVector", "ov", ox, oy, oz);
    addAttribute(outVec);

    //Attribute affects.
    attributeAffects(startMat, outVec);
    attributeAffects(startMat, ox);
    attributeAffects(startMat, oy);
    attributeAffects(startMat, oz);
    attributeAffects(startMat, mag);

    attributeAffects(endMat, outVec);
    attributeAffects(endMat, ox);
    attributeAffects(endMat, oy);
    attributeAffects(endMat, oz);
    attributeAffects(endMat, mag);

    attributeAffects(normAttr, outVec);
    attributeAffects(normAttr, ox);
    attributeAffects(normAttr, oy);
    attributeAffects(normAttr, oz);
    attributeAffects(normAttr, mag);

    return MS::kSuccess;

}

File: Python/cdVector.py
import sys
import maya.OpenMaya as OpenMaya
import maya.OpenMayaMPx as OpenMayaMPx

kPluginNodeTypeName = 'cdVector'
kPluginNodeTypeId = OpenMaya.MTypeId(0x123e1)


class cdVector(OpenMayaMPx.MPxNode):

    startMat = OpenMaya.MObject()  # World Matrix of our starting obj.
    endMat = OpenMaya.MObject()    # World Matrix of our ending obj.
    mag = OpenMaya.MObject()       # Magnitude of vector.
    normAttr = OpenMaya.MObject()  # Normalize boolean.
    outVec = OpenMaya.MObject()    # Output vector.
    ox = OpenMaya.MObject()        # X value of vector.
    oy = OpenMaya.MObject()        # Y value of vector.
    oz = OpenMaya.MObject()        # Z value of vector.

    def __init__(self):
        OpenMayaMPx.MPxNode.__init__(self)

    def compute(self, plug, dataBlock):

        if plug == cdVector.mag or\
           plug == cdVector.outVec or\
           plug == cdVector.ox or\
           plug == cdVector.oy or\
           plug == cdVector.oz:

            #Grab out incoming plugs.
            startMat = dataBlock.inputValue(cdVector.startMat).asMatrix()
            endMat = dataBlock.inputValue(cdVector.endMat).asMatrix()
            norm = dataBlock.inputValue(cdVector.normAttr).asBool()

            startVec = OpenMaya.MVector(startMat(3, 0),
                                        startMat(3, 1),
                                        startMat(3, 2))

            endVec = OpenMaya.MVector(endMat(3, 0),
                                      endMat(3, 1),
                                      endMat(3, 2))

            upVec = endVec - startVec

            dist = upVec.length()

            outVecX = dataBlock.outputValue(cdVector.ox)
            outVecY = dataBlock.outputValue(cdVector.oy)
            outVecZ = dataBlock.outputValue(cdVector.oz)
            magVal = dataBlock.outputValue(cdVector.mag)

            if dist > .001:

                if norm:
                    upVec.normalize()
                    outVecX.setFloat(upVec.x)
                    outVecY.setFloat(upVec.y)
                    outVecZ.setFloat(upVec.z)
                    magVal.setFloat(1.0)

                else:
                    outVecX.setFloat(upVec.x)
                    outVecY.setFloat(upVec.y)
                    outVecZ.setFloat(upVec.z)
                    magVal.setFloat(dist)

                dataBlock.setClean(plug)
            else:
                outVecX.setFloat(0.0)
                outVecY.setFloat(0.0)
                outVecZ.setFloat(0.0)


def nodeCreator():
    return OpenMayaMPx.asMPxPtr(cdVector())


def nodeInitializer():
    nAttr = OpenMaya.MFnNumericAttribute()
    mAttr = OpenMaya.MFnMatrixAttribute()

    cdVector.startMat = mAttr.create("startMatrix",
                                     "startMat",
                                     OpenMaya.MFnMatrixAttribute.kDouble)

    cdVector.addAttribute(cdVector.startMat)

    cdVector.endMat = mAttr.create("endMatrix",
                                   "endMat",
                                   OpenMaya.MFnMatrixAttribute.kDouble)

    cdVector.addAttribute(cdVector.endMat)

    cdVector.mag = nAttr.create("magnitude",
                                "mag",
                                OpenMaya.MFnNumericData.kFloat,
                                0.0)
    cdVector.addAttribute(cdVector.mag)

    cdVector.normAttr = nAttr.create("normalize",
                                     "norm",
                                     OpenMaya.MFnNumericData.kBoolean)

    cdVector.addAttribute(cdVector.normAttr)

    #Output Rotation attribute.
    cdVector.ox = nAttr.create("outVectorX",
                               "ox",
                               OpenMaya.MFnNumericData.kFloat,
                               0.0)
    nAttr.setWritable(False)
    nAttr.setStorable(True)
    nAttr.setReadable(True)

    cdVector.oy = nAttr.create("outVectorY",
                               "oy",
                               OpenMaya.MFnNumericData.kFloat,
                               0.0)
    nAttr.setWritable(False)
    nAttr.setStorable(True)
    nAttr.setReadable(True)

    cdVector.oz = nAttr.create("outVectorZ",
                               "oz",
                               OpenMaya.MFnNumericData.kFloat,
                               0.0)
    nAttr.setWritable(False)
    nAttr.setStorable(True)
    nAttr.setReadable(True)

    cdVector.outVec = nAttr.create("outputVector",
                                   "ov",
                                   cdVector.ox,
                                   cdVector.oy,
                                   cdVector.oz)

    cdVector.addAttribute(cdVector.outVec)

    #Attribute affects.
    cdVector.attributeAffects(cdVector.startMat, cdVector.outVec)
    cdVector.attributeAffects(cdVector.startMat, cdVector.ox)
    cdVector.attributeAffects(cdVector.startMat, cdVector.oy)
    cdVector.attributeAffects(cdVector.startMat, cdVector.oz)
    cdVector.attributeAffects(cdVector.startMat, cdVector.mag)

    cdVector.attributeAffects(cdVector.endMat, cdVector.outVec)
    cdVector.attributeAffects(cdVector.endMat, cdVector.ox)
    cdVector.attributeAffects(cdVector.endMat, cdVector.oy)
    cdVector.attributeAffects(cdVector.endMat, cdVector.oz)
    cdVector.attributeAffects(cdVector.endMat, cdVector.mag)

    cdVector.attributeAffects(cdVector.normAttr, cdVector.outVec)
    cdVector.attributeAffects(cdVector.normAttr, cdVector.ox)
    cdVector.attributeAffects(cdVector.normAttr, cdVector.oy)
    cdVector.attributeAffects(cdVector.normAttr, cdVector.oz)
    cdVector.attributeAffects(cdVector.normAttr, cdVector.mag)


def initializePlugin(mobject):
    '''
    Initializes the plugin.

    Returns: None.
    '''

    mplugin = OpenMayaMPx.MFnPlugin(mobject, "Chris DeVito", "1.0", "Any")
    try:
        mplugin.registerNode(kPluginNodeTypeName,
                             kPluginNodeTypeId,
                             nodeCreator,
                             nodeInitializer,
                             OpenMayaMPx.MPxNode.kDependNode)

        sys.stderr.write("Registered node: %s\n" % kPluginNodeTypeName)
    except:
        sys.stderr.write("Failed to register node: %s\n" % kPluginNodeTypeName)


def uninitializePlugin(mobject):
    '''
    Uninitializes the plugin.

    Returns: None.
    '''

    mplugin = OpenMayaMPx.MFnPlugin(mobject)
    try:
        mplugin.deregisterNode(kPluginNodeTypeId)
        sys.stderr.write("Deregistered node: %s\n" % kPluginNodeTypeName)
    except:
        sys.stderr.write("Failed to deregister node: %s\n"
                         % kPluginNodeTypeName)
